### ISSUE TITLE
"429 too many outstanding requests" occurred even when there're only few requests

### ISSUE THREAD

**Issue by suraciii** [grafana/tempo#2315]
**Describe the bug**

429 error when finding trace by id.

> failed to get trace with id: 0000000000000000fdc76854a6965840 Status: 500 Internal Server Error Body: 3 errors occurred: * rpc error: code = Code(429) desc = too many outstanding requests * rpc error: code = Code(429) desc = too many outstanding requests * rpc error: code = Code(429) desc = too many outstanding requests

**To Reproduce**
1. Start Tempo (2.0.0)
2. Perform Operations (Find Trace By Id)

**Expected behavior**
Request complete successfully

**Environment:**
 - Infrastructure: [Kubernetes]
 - Deployment tool: [custom]

**Additional Context**

configs:
```yaml
query_frontend:
    max_outstanding_per_tenant: 8000
    querier_forget_delay: 0s
    max_retries: 2
    search:
        concurrent_jobs: 2000
        target_bytes_per_job: 10485760
        default_result_limit: 20
        max_result_limit: 0
        max_duration: 168h0m0s
        query_backend_after: 15m0s
        query_ingesters_until: 30m0s
    trace_by_id:
        query_shards: 80
        hedge_requests_at: 2s
        hedge_requests_up_to: 3
```

metrics:

<img width="1199" alt="image" src="https://user-images.githubusercontent.com/13304483/230823854-0c8db483-e533-48c9-ac6d-a92143acc2db.png">

<img width="1195" alt="image" src="https://user-images.githubusercontent.com/13304483/230823812-ac8ad904-19ff-4649-8a0d-59ad1133ee5a.png">

<img width="1206" alt="image" src="https://user-images.githubusercontent.com/13304483/230824028-3a9bcd83-595e-4e73-a709-4bbd8cb62e4d.png">

**Comment by mapno** [comment]
Hi @suraciii!

`too many outstanding requests` can also be returned by the queue that's used for passing search jobs from the query-frontend and the querier.

When a request is received in the query-frontend, it gets split into many smaller jobs. These jobs are pushed concurrently into a queue (up to `query_frontend.search. concurrent_jobs`), which has a max configured via `query_frontend.max_outstanding_per_tenant`.

Based on your config, if the search requests are enormous, 4-5 of them could fill the entire queue and throw that error you're seeing (it's probably more in practice).

**Comment by suraciii** [comment]
Thanks for responding.

@mapno 
> too many outstanding requests can also be returned by the queue that's used for passing search jobs from the query-frontend and the querier.
We're not using the searching function of tempo, never a search job requested.

@joe-elliott 
> How many queriers are you running?
7 querier pods and 2 query-frontend pods.

> Can you share your querier config?
```yaml
querier:
    search: # search function are not used
        query_timeout: 1m0s
        prefer_self: 0
        external_endpoints:
            - http://192.168.22.57:8080
        external_hedge_requests_at: 8s
        external_hedge_requests_up_to: 2
    trace_by_id:
        query_timeout: 30s
    max_concurrent_queries: 20
    frontend_worker:
        frontend_address: tempo-query-frontend-discovery.tempo.svc.cluster.local:9095
        dns_lookup_duration: 10s
        parallelism: 2
        match_max_concurrent: true
        id: ""
        grpc_client_config:
            max_recv_msg_size: 104857600
            max_send_msg_size: 16777216
            grpc_compression: gzip
            rate_limit: 0
            rate_limit_burst: 0
            backoff_on_ratelimits: false
            backoff_config:
                min_period: 100ms
                max_period: 1s
                max_retries: 5
            tls_enabled: false
            tls_cert_path: ""
            tls_key_path: ""
            tls_ca_path: ""
            tls_server_name: ""
            tls_insecure_skip_verify: false
            tls_cipher_suites: ""
            tls_min_version: ""
    query_relevant_ingesters: false
```

> The value of the cortex_query_frontend_queue_length metric
<img width="1188" alt="image" src="https://user-images.githubusercontent.com/13304483/231044496-a124e937-7b05-4c9f-931a-6b1573ed666b.png">

**Comment by joe-elliott** [comment]
That is really strange. It looks like your queue is not being drawn down at all. Are any queries being processed?

**Comment by suraciii** [comment]
I increased max_outstanding_per_tenant to 20000 and the issue has never occured since then:

<img width="1183" alt="image" src="https://user-images.githubusercontent.com/13304483/234245702-7090f141-b27b-4507-8519-8a0df2d55aaa.png">
<img width="1189" alt="image" src="https://user-images.githubusercontent.com/13304483/234246762-e53802cd-10f6-4bf0-995a-cb98ac0ed82f.png">
<img width="1194" alt="image" src="https://user-images.githubusercontent.com/13304483/234246885-8864a82b-7411-479d-9707-7a5e394d18f4.png">

What confused me are:
1. The previous value of max_outstanding_per_tenant was 8000, but the metrics show that no queue length of each pod reached that value.
2. As you pointed out, the queue length did not decrease until I restarted the query-frontend pod, so once a pod ran into this issue, it could not recover (in time, I guess).
